docs/examples/nmt/gnmt.md

```diff
@@ -0,0 +1,438 @@
+
+```python
+def cache_dataset(dataset, prefix):
+    """Cache the processed npy dataset  the dataset into a npz
+
+    Parameters
+    ----------
+    dataset : SimpleDataset
+    file_path : str
+    """
+    if not os.path.exists(_C.CACHE_PATH):
+        os.makedirs(_C.CACHE_PATH)
+    src_data = np.array([ele[0] for ele in dataset])
+    tgt_data = np.array([ele[1] for ele in dataset])
+    np.savez(os.path.join(_C.CACHE_PATH, prefix + '.npz'), src_data=src_data, tgt_data=tgt_data)
+
+
+def load_cached_dataset(prefix):
+    cached_file_path = os.path.join(_C.CACHE_PATH, prefix + '.npz')
+    if os.path.exists(cached_file_path):
+        print('Load cached data from {}'.format(cached_file_path))
+        dat = np.load(cached_file_path)
+        return ArrayDataset(np.array(dat['src_data']), np.array(dat['tgt_data']))
+    else:
+        return None
+
+
+class TrainValDataTransform(object):
+    """Transform the machine translation dataset.
+
+    Clip source and the target sentences to the maximum length. For the source sentence, append the
+    EOS. For the target sentence, append BOS and EOS.
+
+    Parameters
+    ----------
+    src_vocab : Vocab
+    tgt_vocab : Vocab
+    src_max_len : int
+    tgt_max_len : int
+    """
+    def __init__(self, src_vocab, tgt_vocab, src_max_len, tgt_max_len):
+        self._src_vocab = src_vocab
+        self._tgt_vocab = tgt_vocab
+        self._src_max_len = src_max_len
+        self._tgt_max_len = tgt_max_len
+
+    def __call__(self, src, tgt):
+        if self._src_max_len > 0:
+            src_sentence = self._src_vocab[src.split()[:self._src_max_len]]
+        else:
+            src_sentence = self._src_vocab[src.split()]
+        if self._tgt_max_len > 0:
+            tgt_sentence = self._tgt_vocab[tgt.split()[:self._tgt_max_len]]
+        else:
+            tgt_sentence = self._tgt_vocab[tgt.split()]
+        src_sentence.append(self._src_vocab[self._src_vocab.eos_token])
+        tgt_sentence.insert(0, self._tgt_vocab[self._tgt_vocab.bos_token])
+        tgt_sentence.append(self._tgt_vocab[self._tgt_vocab.eos_token])
+        src_npy = np.array(src_sentence, dtype=np.int32)
+        tgt_npy = np.array(tgt_sentence, dtype=np.int32)
+        return src_npy, tgt_npy
+
+
+def process_dataset(dataset, src_vocab, tgt_vocab, src_max_len=-1, tgt_max_len=-1):
+    start = time.time()
+    dataset_processed = dataset.transform(TrainValDataTransform(src_vocab, tgt_vocab,
+                                                                src_max_len,
+                                                                tgt_max_len), lazy=False)
+    end = time.time()
+    print('Processing Time spent: {}'.format(end - start))
+    return dataset_processed
+
+
+def load_translation_data(dataset, src_lang='en', tgt_lang='vi'):
+    """Load translation dataset
+
+    Parameters
+    ----------
+    dataset : str
+    src_lang : str, default 'en'
+    tgt_lang : str, default 'vi'
+
+    Returns
+    -------
+
```

Comment:
returns ..?

Reply:
We use capital one in gluonnlp.

Reply:
I meant the return types are missing

Reply:
@szhengac @sxjscience seems like the return types are still missing.